In [114]:
import pandas as pd
import numpy as np

In [115]:
df=pd.read_csv("amazon_delivery.csv")

In [116]:
df.head(5)

Order_ID  Agent_Age  Agent_Rating  Store_Latitude  Store_Longitude  \
0  ialx566343618         37           4.9       22.745049        75.892471   
1  akqg208421122         34           4.5       12.913041        77.683237   
2  njpu434582536         23           4.4       12.914264        77.678400   
3  rjto796129700         38           4.7       11.003669        76.976494   
4  zguw716275638         32           4.6       12.972793        80.249982   

   Drop_Latitude  Drop_Longitude  Order_Date Order_Time Pickup_Time  \
0      22.765049       75.912471  2022-03-19   11:30:00    11:45:00   
1      13.043041       77.813237  2022-03-25   19:45:00    19:50:00   
2      12.924264       77.688400  2022-03-19   08:30:00    08:45:00   
3      11.053669       77.026494  2022-04-05   18:00:00    18:10:00   
4      13.012793       80.289982  2022-03-26   13:30:00    13:45:00   

      Weather  Traffic      Vehicle            Area  Delivery_Time  \
0       Sunny    High   motorcycle           Urban             120   
1      Stormy     Jam      scooter   Metropolitian             165   
2  Sandstorms     Low   motorcycle           Urban             130   
3       Sunny  Medium   motorcycle   Metropolitian             105   
4      Cloudy    High      scooter   Metropolitian             150   

      Category  
0     Clothing  
1  Electronics  
2       Sports  
3    Cosmetics  
4         Toys

In [117]:
df.columns

Index(['Order_ID', 'Agent_Age', 'Agent_Rating', 'Store_Latitude',
       'Store_Longitude', 'Drop_Latitude', 'Drop_Longitude', 'Order_Date',
       'Order_Time', 'Pickup_Time', 'Weather', 'Traffic', 'Vehicle', 'Area',
       'Delivery_Time', 'Category'],
      dtype='object')

Hypothesis Testing Questions
Agent Age vs Delivery Time:Agent Age vs Delivery Time:

# Is there a significant difference in average delivery times between younger and older delivery agents?

# Hypothesis:
Null Hypothesis (H0): There is no significant difference in delivery times between younger and older delivery agents.
Alternative Hypothesis (H1): There is a significant difference in delivery times between younger and older delivery agents.

In [118]:
df.dtypes

Order_ID            object
Agent_Age            int64
Agent_Rating       float64
Store_Latitude     float64
Store_Longitude    float64
Drop_Latitude      float64
Drop_Longitude     float64
Order_Date          object
Order_Time          object
Pickup_Time         object
Weather             object
Traffic             object
Vehicle             object
Area                object
Delivery_Time        int64
Category            object
dtype: object

In [119]:
#Data Cleaning
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43739 entries, 0 to 43738
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Order_ID         43739 non-null  object 
 1   Agent_Age        43739 non-null  int64  
 2   Agent_Rating     43685 non-null  float64
 3   Store_Latitude   43739 non-null  float64
 4   Store_Longitude  43739 non-null  float64
 5   Drop_Latitude    43739 non-null  float64
 6   Drop_Longitude   43739 non-null  float64
 7   Order_Date       43739 non-null  object 
 8   Order_Time       43739 non-null  object 
 9   Pickup_Time      43739 non-null  object 
 10  Weather          43648 non-null  object 
 11  Traffic          43739 non-null  object 
 12  Vehicle          43739 non-null  object 
 13  Area             43739 non-null  object 
 14  Delivery_Time    43739 non-null  int64  
 15  Category         43739 non-null  object 
dtypes: float64(5), int64(2), object(9)
memory usage: 5.3+ MB


In [120]:
df.describe()

Agent_Age  Agent_Rating  Store_Latitude  Store_Longitude  \
count  43739.000000  43685.000000    43739.000000     43739.000000   
mean      29.567137      4.633780       17.210960        70.661177   
std        5.815155      0.334716        7.764225        21.475005   
min       15.000000      1.000000      -30.902872       -88.366217   
25%       25.000000      4.500000       12.933298        73.170283   
50%       30.000000      4.700000       18.551440        75.898497   
75%       35.000000      4.900000       22.732225        78.045359   
max       50.000000      6.000000       30.914057        88.433452   

       Drop_Latitude  Drop_Longitude  Delivery_Time  
count   43739.000000    43739.000000   43739.000000  
mean       17.459031       70.821842     124.905645  
std         7.342950       21.153148      51.915451  
min         0.010000        0.010000      10.000000  
25%        12.985996       73.280000      90.000000  
50%        18.633626       76.002574     125.000000  
75%        22.785049       78.104095     160.000000  
max        31.054057       88.563452     270.000000

# Agent_Age and Insight:
Count: 43,739
Mean: 29.57 years
Standard Deviation: 5.82 years
Min: 15 years
25th Percentile: 25 years
50th Percentile (Median): 30 years
75th Percentile: 35 years
Max: 50 years
Insights:
The average age of delivery agents is around 29.57 years, with most agents being between 25 and 35 years old.
There are some very young agents (15 years old), which might need further investigation for validity.

# Agent_Rating and insight:
Count: 43,685
Mean: 4.63
Standard Deviation: 0.33
Min: 1.00
25th Percentile: 4.50
50th Percentile (Median): 4.70
75th Percentile: 4.90
Max: 6.00
Insights:

The average rating is quite high at 4.63, indicating generally positive feedback for agents.
The ratings range from 1 to 6, with most agents scoring between 4.5 and 4.9.
There are outlier ratings (1.00 and 6.00) that might require further examination.

Insights from Percentiles:
The ratings are generally high, with half of the agents scoring at least 4.70.
Most of the agents (75%) have ratings between 4.50 and 4.90, indicating a tight cluster of relatively high ratings.
The fact that the 75th percentile is very close to the maximum rating (4.90 compared to 6.00) suggests that most agents perform well, with only a small fraction receiving the very highest or lowest scores.

In [121]:
df

Order_ID  Agent_Age  Agent_Rating  Store_Latitude  \
0      ialx566343618         37           4.9       22.745049   
1      akqg208421122         34           4.5       12.913041   
2      njpu434582536         23           4.4       12.914264   
3      rjto796129700         38           4.7       11.003669   
4      zguw716275638         32           4.6       12.972793   
...              ...        ...           ...             ...   
43734  jlxf819993117         30           4.8       26.902328   
43735  aevx342135787         21           4.6        0.000000   
43736  xnek760674819         30           4.9       13.022394   
43737  cynl434665991         20           4.7       11.001753   
43738  nsyz997960170         23           4.9       23.351058   

       Store_Longitude  Drop_Latitude  Drop_Longitude  Order_Date Order_Time  \
0            75.892471      22.765049       75.912471  2022-03-19   11:30:00   
1            77.683237      13.043041       77.813237  2022-03-25   19:45:00   
2            77.678400      12.924264       77.688400  2022-03-19   08:30:00   
3            76.976494      11.053669       77.026494  2022-04-05   18:00:00   
4            80.249982      13.012793       80.289982  2022-03-26   13:30:00   
...                ...            ...             ...         ...        ...   
43734        75.794257      26.912328       75.804257  2022-03-24   11:35:00   
43735         0.000000       0.070000        0.070000  2022-02-16   19:55:00   
43736        80.242439      13.052394       80.272439  2022-03-11   23:50:00   
43737        76.986241      11.041753       77.026241  2022-03-07   13:35:00   
43738        85.325731      23.431058       85.405731  2022-03-02   17:10:00   

      Pickup_Time     Weather  Traffic      Vehicle            Area  \
0        11:45:00       Sunny    High   motorcycle           Urban    
1        19:50:00      Stormy     Jam      scooter   Metropolitian    
2        08:45:00  Sandstorms     Low   motorcycle           Urban    
3        18:10:00       Sunny  Medium   motorcycle   Metropolitian    
4        13:45:00      Cloudy    High      scooter   Metropolitian    
...           ...         ...      ...          ...             ...   
43734    11:45:00       Windy    High   motorcycle   Metropolitian    
43735    20:10:00       Windy     Jam   motorcycle   Metropolitian    
43736    00:05:00      Cloudy     Low      scooter   Metropolitian    
43737    13:40:00      Cloudy    High   motorcycle   Metropolitian    
43738    17:15:00         Fog  Medium      scooter   Metropolitian    

       Delivery_Time     Category  
0                120     Clothing  
1                165  Electronics  
2                130       Sports  
3                105    Cosmetics  
4                150         Toys  
...              ...          ...  
43734            160         Home  
43735            180      Jewelry  
43736             80         Home  
43737            130      Kitchen  
43738            180    Cosmetics  

[43739 rows x 16 columns]

In [122]:
df.dtypes

Order_ID            object
Agent_Age            int64
Agent_Rating       float64
Store_Latitude     float64
Store_Longitude    float64
Drop_Latitude      float64
Drop_Longitude     float64
Order_Date          object
Order_Time          object
Pickup_Time         object
Weather             object
Traffic             object
Vehicle             object
Area                object
Delivery_Time        int64
Category            object
dtype: object

In [123]:
# Convert Order_Date to datetime
df['Order_Date'] = pd.to_datetime(df['Order_Date'])



In [124]:
df.dtypes

Order_ID                   object
Agent_Age                   int64
Agent_Rating              float64
Store_Latitude            float64
Store_Longitude           float64
Drop_Latitude             float64
Drop_Longitude            float64
Order_Date         datetime64[ns]
Order_Time                 object
Pickup_Time                object
Weather                    object
Traffic                    object
Vehicle                    object
Area                       object
Delivery_Time               int64
Category                   object
dtype: object

In [125]:
df.head()

Order_ID  Agent_Age  Agent_Rating  Store_Latitude  Store_Longitude  \
0  ialx566343618         37           4.9       22.745049        75.892471   
1  akqg208421122         34           4.5       12.913041        77.683237   
2  njpu434582536         23           4.4       12.914264        77.678400   
3  rjto796129700         38           4.7       11.003669        76.976494   
4  zguw716275638         32           4.6       12.972793        80.249982   

   Drop_Latitude  Drop_Longitude Order_Date Order_Time Pickup_Time  \
0      22.765049       75.912471 2022-03-19   11:30:00    11:45:00   
1      13.043041       77.813237 2022-03-25   19:45:00    19:50:00   
2      12.924264       77.688400 2022-03-19   08:30:00    08:45:00   
3      11.053669       77.026494 2022-04-05   18:00:00    18:10:00   
4      13.012793       80.289982 2022-03-26   13:30:00    13:45:00   

      Weather  Traffic      Vehicle            Area  Delivery_Time  \
0       Sunny    High   motorcycle           Urban             120   
1      Stormy     Jam      scooter   Metropolitian             165   
2  Sandstorms     Low   motorcycle           Urban             130   
3       Sunny  Medium   motorcycle   Metropolitian             105   
4      Cloudy    High      scooter   Metropolitian             150   

      Category  
0     Clothing  
1  Electronics  
2       Sports  
3    Cosmetics  
4         Toys

In [126]:
df.dtypes

Order_ID                   object
Agent_Age                   int64
Agent_Rating              float64
Store_Latitude            float64
Store_Longitude           float64
Drop_Latitude             float64
Drop_Longitude            float64
Order_Date         datetime64[ns]
Order_Time                 object
Pickup_Time                object
Weather                    object
Traffic                    object
Vehicle                    object
Area                       object
Delivery_Time               int64
Category                   object
dtype: object

In [127]:
# Replace 'NaT' with actual NaN values and convert to string
df['Order_Time'] = df['Order_Time'].replace('NaT', pd.NaT).astype(str)
df['Pickup_Time'] = df['Pickup_Time'].replace('NaT', pd.NaT).astype(str)

# Ensure all entries are strings before stripping
df['Order_Time'] = df['Order_Time'].apply(lambda x: str(x).strip())
df['Pickup_Time'] = df['Pickup_Time'].apply(lambda x: str(x).strip())

# Remove any internal whitespace and ensure correct format
df['Order_Time'] = df['Order_Time'].apply(lambda x: ' '.join(x.split()) if x != 'nan' else x)
df['Pickup_Time'] = df['Pickup_Time'].apply(lambda x: ' '.join(x.split()) if x != 'nan' else x)

# Convert to datetime.time, with errors='coerce' to handle any invalid formats
df['Order_Time'] = pd.to_datetime(df['Order_Time'], format='%H:%M:%S', errors='coerce').dt.time
df['Pickup_Time'] = pd.to_datetime(df['Pickup_Time'], format='%H:%M:%S', errors='coerce').dt.time

print(df)

            Order_ID  Agent_Age  Agent_Rating  Store_Latitude  \
0      ialx566343618         37           4.9       22.745049   
1      akqg208421122         34           4.5       12.913041   
2      njpu434582536         23           4.4       12.914264   
3      rjto796129700         38           4.7       11.003669   
4      zguw716275638         32           4.6       12.972793   
...              ...        ...           ...             ...   
43734  jlxf819993117         30           4.8       26.902328   
43735  aevx342135787         21           4.6        0.000000   
43736  xnek760674819         30           4.9       13.022394   
43737  cynl434665991         20           4.7       11.001753   
43738  nsyz997960170         23           4.9       23.351058   

       Store_Longitude  Drop_Latitude  Drop_Longitude Order_Date Order_Time  \
0            75.892471      22.765049       75.912471 2022-03-19   11:30:00   
1            77.683237      13.043041       77.813237 2022-03

# Hypothesis Testing:



In [128]:
Thrushold=30
youngeragent=df[df['Agent_Age']<Thrushold]['Delivery_Time']
olderagent=df[df['Agent_Age']>=Thrushold]['Delivery_Time']
print("Younger agent mean delivery time",youngeragent.mean())
print("Older agent mean delivery time",olderagent.mean())

Younger agent mean delivery time 109.28980759470309
Older agent mean delivery time 140.24336082661108


In [129]:
#t-test is a statistical test used to compare the means of two groups
from scipy import stats
t_stat, p_value = stats.ttest_ind(youngeragent,olderagent)

In [130]:
alpha = 0.05
if p_value < alpha:
    print("Reject the null hypothesis. There is a significant difference in delivery times between younger and older delivery agents.")
else:
    print("Fail to reject the null hypothesis. There is no significant difference in delivery times between younger and older delivery agents.")

Reject the null hypothesis. There is a significant difference in delivery times between younger and older delivery agents.


# Rejecting the null hypothesis. There is a significant difference in delivery times between younger and older delivery agents.






# Regression Analysis

In [131]:
#What are the key factors that influence delivery time?

Approach: Perform a multiple linear regression analysis with Delivery_Time as the dependent variable and Agent_Age, Agent_Rating, Weather, Traffic, Vehicle, Area, Store_Latitude, Store_Longitude, Drop_Latitude, Drop_Longitude as independent variables.

In [132]:
# Ensure that the columns are numeric
df['Delivery_Time'] = pd.to_numeric(df['Delivery_Time'], errors='coerce')
df['Agent_Age'] = pd.to_numeric(df['Agent_Age'], errors='coerce')

# Drop rows with any NaN values
df.dropna(subset=['Delivery_Time', 'Agent_Age'], inplace=True)

# Define independent and dependent variables
X = df[['Agent_Age']]
y = df['Delivery_Time']

# Add a constant to the independent variable (for the intercept)
X = sm.add_constant(X)

# Fit the regression model
model = sm.OLS(y, X).fit()

# Print the summary
print(model.summary())


                            OLS Regression Results                            
Dep. Variable:          Delivery_Time   R-squared:                       0.065
Model:                            OLS   Adj. R-squared:                  0.065
Method:                 Least Squares   F-statistic:                     3032.
Date:                Mon, 29 Jul 2024   Prob (F-statistic):               0.00
Time:                        19:22:10   Log-Likelihood:            -2.3335e+05
No. Observations:               43739   AIC:                         4.667e+05
Df Residuals:                   43737   BIC:                         4.667e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         57.6956      1.244     46.381      0.0

The regression model suggests that there is a statistically significant relationship between the age of delivery agents and delivery times, with older agents tending to have longer delivery times. However, the low R-squared value indicates that age alone is not a strong predictor of delivery time, and other factors may also be important.

Time Series Analysis
Trends in Delivery Times Over Time:

Question: What are the trends in delivery times over different months or years?
Approach: Perform a time series analysis on Delivery_Time using Order_Date to identify trends, seasonality, and any cyclic patterns.
Seasonal Impact on Delivery Performance:

Question: How does delivery performance vary across different seasons of the year?
Approach: Perform a seasonal decomposition of Delivery_Time to analyze how delivery performance changes with seasons.

In [134]:
df['Order_Date'] = pd.to_datetime(df['Order_Date'])
df.set_index('Order_Date', inplace=True)

period =7  # Adjust this value based on your data
result = seasonal_decompose(df['Delivery_Time'], model='additive', period=period)

# Plot the decomposition
result.plot()


KeyError: 'Order_Date'